### ISSUE TITLE
app.py

### ISSUE THREAD

**Issue by prernatanksalkar** [prernatanksalkar/FMCG-price-prediction-ML-project#1]
import streamlit as st
import pandas as pd
import joblib
import matplotlib.pyplot as plt
import seaborn as sns

# ==============================================
# 🔹 Load Trained Models and Columns
# ==============================================
rf_model = joblib.load("random_forest_model.pkl")
xgb_model = joblib.load("xgb_model.pkl")
lgbm_model = joblib.load("lgbm_model.pkl")
X_columns = joblib.load("X_columns.pkl")

# ==============================================
# 🎨 Streamlit Page Configuration
# ==============================================
st.set_page_config(
    page_title="Smart Price Predictor",
    page_icon="💰",
    layout="wide",
    initial_sidebar_state="expanded"
)

st.title("💸 Smart Selling Price Predictor for FMCG Products")
st.markdown("""
This app predicts the **selling price** of grocery items based on various product attributes 
like cost, discount, rating, and category — using **Machine Learning Models** 🤖  
""")

# ==============================================
# 📊 Sidebar — User Inputs
# ==============================================
st.sidebar.header("📋 Enter Product Details")

ratings = st.sidebar.number_input("⭐ Product Rating", min_value=0.0, max_value=5.0, value=4.2, step=0.1)
bought_ratio = st.sidebar.number_input("🛍️ Bought Ratio", min_value=0, value=2000)
saving_percent = st.sidebar.number_input("💰 Saving %", min_value=0.0, max_value=100.0, value=25.0)
per_kg_price = st.sidebar.number_input("📦 Per Kg Price (₹)", min_value=0.0, value=120.0)
cost_price = st.sidebar.number_input("💵 Cost Price (₹)", min_value=0.0, value=200.0)
discount_percent = st.sidebar.number_input("🎯 Discount Percentage", min_value=0.0, max_value=100.0, value=25.0)
category = st.sidebar.selectbox("🏷️ Product Category", ["Rice", "Wheat", "paneer", "milk", "biscuits", "colddrinks","chips" , "noodles", "oats", "ghee","butter", "besan", "jeera", "sabudana","pasta", "bread", " tomato ketchup", "chocospread", "haldi powder", "salt", "cashew", "poha", "rava", "peanuts", "cheese","honay","Oil", "Other"])

model_choice = st.sidebar.radio("🧠 Choose a Model", ["Random Forest", "XGBoost", "LightGBM"])

# ==============================================
# 🧮 Prepare Input Data
# ==============================================
input_data = pd.DataFrame([{
    'ratings': ratings,
    'Bought ratio': bought_ratio,
    'saving %': saving_percent,
    'per kg price': per_kg_price,
    'cost price': cost_price,
    'discount percentage': discount_percent,
    'category': category
}])

# One-Hot Encode the Category (aligning with training columns)
input_encoded = pd.get_dummies(input_data, columns=['category'])
input_encoded = input_encoded.reindex(columns=X_columns, fill_value=0)

# ==============================================
# 🚀 Model Prediction
# ==============================================
if st.sidebar.button("🔮 Predict Selling Price"):
    if model_choice == "Random Forest":
        model = rf_model
        model_name = "Random Forest Regressor"
    elif model_choice == "XGBoost":
        model = xgb_model
        model_name = "XGBoost Regressor"
    else:
        model = lgbm_model
        model_name = "LightGBM Regressor"

    predicted_price = model.predict(input_encoded)[0]

    # Display Result
    st.subheader("🎯 Predicted Selling Price")
    st.success(f"💰 Estimated Price: ₹{predicted_price:.2f}")
    st.caption(f"Model Used: **{model_name}**")

    # ==============================================
    # 📈 Visual Insights
    # ==============================================

    col1, col2 = st.columns(2)

    with col1:
        st.markdown("#### Feature Importance (How model made its decision)")
        if hasattr(model, "feature_importances_"):
            importance = pd.Series(model.feature_importances_, index=X_columns)
            top_features = importance.sort_values(ascending=False).head(10)
            fig, ax = plt.subplots(figsize=(6, 4))
            sns.barplot(x=top_features.values, y=top_features.index, palette="coolwarm")
            plt.title("Top 10 Important Features")
            st.pyplot(fig)
        else:
            st.info("Feature importance not available for this model.")

    with col2:
        st.markdown("#### Input Summary")
        st.dataframe(input_data.style.highlight_max(axis=0, color="lightgreen"))

    # Comparison chart for model performance
    st.markdown("### 📊 Model Comparison (based on validation results)")
    results = {
        'Model': ['Random Forest', 'XGBoost', 'LightGBM'],
        'R² Score': [0.92, 0.94, 0.93],
        'MAE': [45.3, 41.8, 42.5],
        'RMSE': [60.1, 55.4, 57.2]
    }
    df_results = pd.DataFrame(results)

    fig2, ax2 = plt.subplots(figsize=(6, 4))
    sns.barplot(data=df_results, x='Model', y='R² Score', palette="crest")
    plt.title("R² Score Comparison of Models")
    st.pyplot(fig2)

# ==============================================
# 📘 Footer
# ==============================================
st.markdown("---")
st.caption("Developed with ❤️ ")
